src/vs/workbench/contrib/extensions/browser/extensionsWorkbenchService.ts

```diff
@@ -1876,10 +1904,33 @@ export class ExtensionsWorkbenchService extends Disposable implements IExtension
 			const galleryExtensions = await this.galleryService.getExtensions(infos, { targetPlatform, compatible: true, productVersion: this.getProductVersion(), updateCheck: true }, CancellationToken.None);
 			if (galleryExtensions.length) {
 				await this.syncInstalledExtensionsWithGallery(galleryExtensions);
+				await this.syncMissingFromGallery(galleryExtensions, infos);
```

Comment:
Probably we should inline the check here itself. You can read all `installed` extensions and update the extension. You can change the type of `installed` to `Extension` for updating.

Reply:
I think I got another better idea. Inline the logic of `syncMissingFromGallery` in `syncInstalledExtensionsWithGallery`. This means while syncing installed extensions with gallery you also update the missing gallery extensions. Logically they belong together so it makes sense to compute them together.

Reply:
Here's why I didn't inline it:

In order to compute `syncMissingFromGallery`, we need to know all of the locally installed extensions (`infos`) that _produced_ the `galleryExtensions`.  

If we don't use `infos` as a filter, then we over-count extensions as 'missing' here:


https://github.com/microsoft/vscode/blob/b216502ec41ad540fe9967941fd248e0c0b099a8/src/vs/workbench/contrib/extensions/browser/extensionsWorkbenchService.ts#L686-L701

There are many places `syncInstalledExtensionsWithGallery` is called with a `GalleryExtension` object that doesn't reflect the entire set of installed extensions.  I don't have any `infos` here to use as the filter. It wouldn't make sense at this callsite to do the sync.

Reply:
TL;DR, when calling `syncMissingFromGallery` the pre-condition is that `galleryExtensions` was produced from a complete set of installed extensions.  There are many places `syncInstalledExtensionsWithGallery()` is called where that is not true.

`syncMissingFromGallery()` works by detecting the _absence_ of a `galleryExtensions` mapping to extensions in `this.local` (seen in the above comment's first code block)

Reply:
> There are many places syncInstalledExtensionsWithGallery is called with a GalleryExtension object that doesn't reflect the entire set of installed extensions. I don't have any infos here to use as the filter. It wouldn't make sense at this callsite to do the sync.

That's right, in that case why not take a flag in `syncInstalledExtensionsWithGallery` whether to update missed gallery extensions or not? IMO `syncInstalledExtensionsWithGallery` also checks missing gallery extensions makes sense and complete.